### PR TITLE
First party data support

### PR DESCRIPTION
**Pull request by RossGoncharuk**
- Implemented server-side support of First Party Data processing;
- Updated Rubicon bidder to process First Party Data fields;
- Updated existing unit tests and added new ones to cover changes.

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java [hunk 1]
```diff
@@ -235,14 +240,41 @@ private boolean isMetricSupported(Metric metric) {
...
+        if (data != null && !data.isNull()) {
+            inventoryObject.setAll(data);
+        }
+
+        if (site != null) {
+            final String search = site.getSearch();
+            if (search != null) {
+                inventoryObject.put("search", search);
+            }
+            return resolveJsonNode(inventoryObject, site.getExt().get("data"));
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java]: Can `ext` be NULL?
Pls check the same for `app`.

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 2]
```diff
@@ -325,27 +336,44 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
...
+                                                   Map<String, String> aliases, List<Imp> imps,
+                                                   Map<Integer, Boolean> vendorsToGdpr) {
 
         final Device device = bidRequest.getDevice();
         final Integer deviceLmt = device != null ? device.getLmt() : null;
         final Map<String, Boolean> bidderToMaskingRequired = bidders.stream()
                 .collect(Collectors.toMap(Function.identity(),
                         bidder -> isMaskingRequiredBidder(vendorsToGdpr, bidder, aliases, deviceLmt)));
 
+        final List<String> allowedBidders = dataBidders(requestExt);
```
[the remark this answers is not in the document]
Comment (RossGoncharuk) [reply src/main/java/org/prebid/server/auction/ExchangeService.java]: I guess we don't even need to pass bidders list to this method, but rather check if our `firstPartyDataBidders` contain the `bidder` from `bidders` when we make `bidderRequests`.

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 3]
```diff
@@ -355,6 +374,79 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
         return bidderRequests;
     }
 
+    /**
+     * Extracts a list of bidders for which first party data is allowed from {@link ExtRequestPrebidData} model.
+     */
+    private static List<String> firstPartyDataBidders(ExtBidRequest requestExt) {
+        final ExtRequestPrebid prebid = requestExt == null ? null : requestExt.getPrebid();
+        final ExtRequestPrebidData data = prebid == null ? null : prebid.getData();
+        return data == null ? Collections.emptyList() : data.getBidders();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Can `data.getBidders()` be NULL ?
Do checking, like: 
```
        final List<String> bidders = data == null ? null : data.getBidders();
        return ObjectUtils.defaultIfNull(bidders, Collections.emptyList());
```

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 4]
```diff
@@ -355,6 +374,79 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
...
+    }
+
+    /**
+     * Checks whether to pass the site.ext.data depending on request having a first party data
+     * allowed for given bidder or not.
+     */
+    private static Site prepareSite(Site site, ExtSite extSite, boolean useFirstPartyData) {
+        final ObjectNode data = extSite == null ? null : extSite.getData();
+
+        return data != null && useFirstPartyData
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Should we invert `useFirstPartyData` ?
Pls check the same for App.

File: src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java [hunk 5]
```diff
@@ -296,24 +331,37 @@ private static RubiconBannerExt makeBannerExt(List<Format> sizes) {
...
+        final boolean hasUser = user != null;
+        final String gender = hasUser ? user.getGender() : null;
+        final Integer yob = hasUser ? user.getYob() : null;
+        final Geo geo = hasUser ? user.getGeo() : null;
         final JsonNode visitor = rubiconImpExt.getVisitor();
-        final RubiconUserExtRp userExtRp = !visitor.isNull() ? RubiconUserExtRp.of(visitor) : null;
+        final boolean hasAnyNotNull = gender != null || yob != null || geo != null || !visitor.isNull();
+
+        final RubiconUserExtRp userExtRp = hasAnyNotNull
+                ? RubiconUserExtRp.of(!visitor.isNull() ? visitor : null, gender, yob, geo) : null;
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java]: Do we need `!visitor.isNull() ? visitor : null` ?

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 6]
```diff
@@ -1452,15 +1461,140 @@ public void shouldSetUserBuyerIdsFromUserExtPrebidAndClearPrebidBuyerIdsAfterwar
...
+    public void shouldPassUserExtDataOnlyForAllowedBidder() {
+        // given
+        final Bidder<?> bidder = mock(Bidder.class);
+        givenBidder("someBidder", bidder, givenEmptySeatBid());
+        givenBidder("missingBidder", bidder, givenEmptySeatBid());
+
+        final ObjectNode dataNode = mapper.createObjectNode();
+        dataNode.put("data", "value");
+
+        final Map<String, Integer> bidderToGdpr = new HashMap<>();
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/auction/ExchangeServiceTest.java]: We can use `doubleMap(...)` for this. Check similar cases.

File: src/test/resources/org/prebid/server/it/auction/rubicon_appnexus/test-auction-rubicon-appnexus-response.json [hunk 7]
```diff
@@ -28,13 +28,13 @@
...
+          "request_body": "{\"id\":\"tid\",\"imp\":[{\"id\":\"adUnitCode3\",\"banner\":{\"format\":[{\"w\":768,\"h\":1024},{\"w\":980,\"h\":400}],\"w\":768,\"h\":1024,\"ext\":{\"rp\":{\"size_id\":80,\"alt_size_ids\":[102],\"mime\":\"text/html\"}}},\"ext\":{\"rp\":{\"zone_id\":4001,\"track\":{\"mint\":\"prebid\",\"mint_version\":\"source1_platform1_version1\"}}}}],\"site\":{\"domain\":\"example.com\",\"page\":\"http://www.example.com\",\"publisher\":{\"ext\":{\"rp\":{\"account_id\":2001}}},\"ext\":{\"rp\":{\"site_id\":3001}}},\"device\":{\"ua\":\"userAgent\",\"dnt\":10,\"ip\":\"192.168.244.1\",\"pxratio\":4.2,\"language\":\"en\",\"ifa\":\"ifaId\",\"ext\":{\"rp\":{\"pixelratio\":4.2}}},\"user\":{\"id\":\"12345\",\"buyeruid\":\"J5VLCWQP-26-CWFT\",\"ext\":{\"rp\":{},\"consent\":\"consent1\"}},\"at\":1,\"tmax\":1000,\"source\":{\"fd\":1,\"tid\":\"tid\"},\"regs\":{\"ext\":{\"gdpr\":1}}}",
```
Comment (rpanchyk) [inline src/test/resources/org/prebid/server/it/auction/rubicon_appnexus/test-auction-rubicon-appnexus-response.json]: Why we obtained `\"rp\":{},` empty object? Can we avoid it?

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 8]
```diff
@@ -355,6 +374,80 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
...
+                : Json.mapper.valueToTree(ExtBidRequest.of(
+                prebid == null ? null : prebid.toBuilder().data(null).build()));
+    }
+
+    /**
+     * Checks whether to pass the site.ext.data depending on request having a first party data
+     * allowed for given bidder or not.
+     */
+    private static Site prepareSite(Site site, ExtSite extSite, boolean useFirstPartyData) {
+        final ObjectNode data = extSite == null ? null : extSite.getData();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Pls rename to `extSiteDataNode`. Do the same for `app`.

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 9]
```diff
@@ -355,6 +374,80 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
...
+
+        return data != null && !useFirstPartyData
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Add `site != null` check here first. Do the same for `app`.

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 10]
```diff
@@ -355,6 +374,80 @@ private boolean isValidBidder(String bidder, Map<String, String> aliases) {
...
+    }
+
+    /**
+     * Checks whether to pass the app.ext.data depending on request having a first party data
+     * allowed for given bidder or not.
+     */
+    private static App prepareApp(App app, ExtApp extApp, boolean useFirstPartyData) {
+        final ObjectNode data = extApp == null ? null : extApp.getData();
+
+        return data != null && !useFirstPartyData
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Same as above.

File: src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java (hunk 5 above)
Comment (rpanchyk) [reply src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java]: yes, we do.

**Pull request merged**